pkg/scrape/scrape.go

```diff
@@ -0,0 +1,193 @@
+package scrape
+
+import (
+	"fmt"
+	"github.com/marcus-crane/khinsider/v2/pkg/util"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/PuerkitoBio/goquery"
+	"github.com/pterm/pterm"
+
+	"github.com/marcus-crane/khinsider/v2/pkg/types"
+)
+
+const (
+	LetterBase = "https://downloads.khinsider.com/game-soundtracks/browse/"
+)
+
+func DownloadPage(url string) (*http.Response, error) {
+	res, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	if res.StatusCode != 200 {
+		return nil, err
+	}
+	return res, err
+}
+
+func GetResultsForLetter(letter string) (types.SearchResults, error) {
+	url := fmt.Sprintf("%s%s", LetterBase, letter)
+	res, err := DownloadPage(url)
+	defer func(Body io.ReadCloser) {
+		err := Body.Close()
+		if err != nil {
+			panic(err)
+		}
+	}(res.Body)
+	if err != nil {
+		return nil, err
+	}
+	doc, err := goquery.NewDocumentFromReader(res.Body)
+	if err != nil {
+		return nil, err
+	}
+	results := make(types.SearchResults)
+	doc.Find("#EchoTopic p[align='left'] a").Each(func(i int, s *goquery.Selection) {
+		title := s.Text()
+		results[title] = "#"
+		trackUrl, exists := s.Attr("href")
+		if exists {
+			results[title] = trackUrl
+		} else {
+			results[title] = "#"
+		}
+	})
+	return results, nil
+}
+
+func RetrieveAlbum(path string) (types.Album, error) {
+	var album types.Album
+	slugBits := strings.Split(path, "/")
+	album.Slug = slugBits[len(slugBits)-1]
+	albumUrl := fmt.Sprintf("%s%s", util.SiteBase, path)
+	res, err := DownloadPage(albumUrl)
+	defer func(Body io.ReadCloser) {
+		err := Body.Close()
+		if err != nil {
+			panic(err)
+		}
+	}(res.Body)
+	if err != nil {
+		return album, err
+	}
+	doc, err := goquery.NewDocumentFromReader(res.Body)
+	if err != nil {
+		return album, err
+	}
+	metadata := doc.Find("#EchoTopic p[align='left'] b")
+	if metadata.Length() == 5 {
+		album.FlacAvailable = true
+	}
+	metadata.Each(func(i int, s *goquery.Selection) {
+		if i == 0 {
+			album.Name = s.Text()
+		}
+		if i == 1 {
+			album.FileCount, err = strconv.Atoi(s.Text())
+			if err != nil {
+				album.FileCount = 0
+			}
+		}
+		if i == 2 {
+			album.MP3FileSize = s.Text()
+		}
+		if i == 3 && album.FlacAvailable {
+			album.FlacFileSize = s.Text()
+		}
+	})
+	flacLabel := ""
+	if album.FlacAvailable {
+		flacLabel = "[FLAC]"
+	}
+	pterm.Success.Printfln(
+		"Found %s (%d tracks) %s %s",
+		album.Name,
+		album.FileCount,
+		"[MP3]",
+		flacLabel)
+	pterm.Warning.Println("Searching for track locations up front. This may take seconds or minutes depending on album length!")
+	_ = doc.Find("#EchoTopic table:not(#songList) tr div a").Each(func(i int, s *goquery.Selection) {
+		coverUrl, exists := s.Attr("href")
+		if exists {
+			// TODO: Use proper escaping. Tried stdlib but it escaped everything
```

Comment:
Does khinsider care?

Reply:
I don't remember anymore but I think there were some covers that were served with spaces which caused me trouble somewhere down the line.